packages/bvaughn-architecture-demo/components/console/MessagesList.tsx

```diff
@@ -0,0 +1,80 @@
+import { Message } from "@replayio/protocol";
+import { useContext } from "react";
+
+import { ConsoleFiltersContext, FocusContext, ReplayClientContext } from "../../src/contexts";
+import useFilteredMessages from "../../src/hooks/useFilteredMessages";
+import { getMessages } from "../../src/suspense/MessagesCache";
+import { getClosestPointForTime } from "../../src/suspense/PointsCache";
+
+import MessageRenderer from "./MessageRenderer";
+import styles from "./MessagesList.module.css";
+
+// This is a crappy approximation of the console; the UI isn't meant to be the focus of this branch.
+// The primary purpose of this component is to showcase:
+// 1. The getMessages() Suspense cache for just-in-time loading of console data
+// 2. The memoized useFilteredMessages() selector for computing derived (merged, sorted, and filtered) data
+//
+// Note that the props passed from the parent component would more likely be exposed through Context in a real app.
+// We're passing them as props in this case since the parent and child are right beside each other in the tree.
+export default function MessagesList() {
+  const replayClient = useContext(ReplayClientContext);
+
+  const { filterByText, levelFlags } = useContext(ConsoleFiltersContext);
+  const { range, isTransitionPending: isFocusTransitionPending } = useContext(FocusContext);
+
+  let focusRange = null;
+  if (range !== null) {
+    const [startTime, endTime] = range;
+
+    const startPoint = getClosestPointForTime(replayClient, startTime);
+    const endPoint = getClosestPointForTime(replayClient, endTime);
```

Comment:
q: if I understand correctly - this is using a simple **waterfall** fetch on render, right? That is certainly fine, especially for demo purposes. I just wonder what's the idiomatic/canonical answer for removing the waterfall but I guess the answer is just to preload both of those and that can be done in parallel so the potential "render waterfall" here wouldn't quite matter. Is that correct understanding of those patterns?

Reply:
I guess that, partially, you have also mentioned the waterfall concern [here](https://github.com/RecordReplay/devtools/pull/6992/files#r885643829) - but based on that helper's API I'm not actually sure how to use it 😅 

It has to be called from render as that is the only place from which one can suspend, right? But I'm surprised that it doesn't return any values cause I've imagined that it could be used like this:
```js
const [startPoint, endPoint] = suspendInParallel([
  () => getClosestPointForTime(replayClient, startTime),
  () => getClosestPointForTime(replayClient, endTime)
])
```

It also seems to "leak" suspense to renders whereas usually it's promoted as something completely hidden from the call site (which is cool!). So perhaps I'm misunderstanding completely how this can be used.

However, I'm still curious if the preloading logic would remove the need for such a helper altogether cause it seems to me that it would but again, I feel like I might miss some details here.

Reply:
The `suspendInParallel` helper could help but doesn't (currently) return the values https://github.com/RecordReplay/devtools/pull/6992#discussion_r885643829

```js
const [startPoint, endPoint] = suspendInParallel([
  getClosestPointForTime(replayClient, startTime),
  getClosestPointForTime(replayClient, endTime),
]);
```

I find it a bit tricky to spot the "suspenders", and by extension its tricky to see where to apply this helper.

Reply:
I have some small follow up changes I'm going to make this morning (actually woke up a few times thinking about them last night 😭) and I'll toss this one in as well 😁

Reply:
TBH the only real snag on `suspendInParallel` is that I don't know how to manage the return type (to be anything other than `any[]`)

Reply:
Oh, I'm your guy then :wink:

```ts
type AnyFunction = (...args: any[]) => any;

declare function suspendInParallel<T extends AnyFunction[]>(
  readers: [...T]
): { [K in keyof T]: ReturnType<Extract<T[K], AnyFunction>> };
```

Note that you might need to cast the final result within the implementation itself because TS won't be able to verify that it matches this annotation. For public consumers this should work great though.

Reply:
Perhaps I should explain this weirdo...

1. TS doesnt really have a tuple constraint (which is super annoying) so we need to use an array constraint on the generic `T extends AnyFunction[]`
2. we need to make it to actually narrow the inline arrays used as arguments here to be of a tuple type and using `[...T]` is a neat trick to do it
3. then we proceed to the return type which is a mapped type - it allows us to iterate over the type per property. And that is why we had to make it infer those array arguments as tuples. We want to "unwrap" types per position. So we just iterate over keys of `T` and then do some type-level manipulation on a specific element (`T[K]`)
4. It should be enough to just use `ReturnType<T[K]>` here and in TS 4.7 it actually is enough... but that's a new improvement. That extra `Extract` just allows us to "cast" this at the generic level and force TS to accept `T[K]` as a type argument passed to `ReturnType` (as it only allows functions as type arguments)

Reply:
Oh hell yes! Thank you for the type info and the explanation :)